### ISSUE TITLE
Avatars stored in attachments directory not moved during upgrade

### ISSUE THREAD

**Issue by Oldiesmann** [SimpleMachines/SMF#5123]
#### Description
In 2.0 and earlier, the default option for avatars is to store them in the attachments directory. This option is no longer available in 2.1, but the avatars are not moved over to the custom_avatars directory during the upgrade, resulting in the avatar not showing for anyone who uploaded one prior to the upgrade.

### Steps to reproduce
1. Upload an avatar in 2.0 with the default settings (store in attachments directory)
2. Upgrade to 2.1

### Environment (complete as necessary)
- Version/Git revision: latest
- Database Type: MaraiDB
- Database Version: 10.0.36
- PHP Version: 7.1.23

### Additional information/references
There is a hidden option in 2.0 (and possibly in 1.1 as well) to move avatars from the attachments directory to a custom avatars directory, so it shouldn't be too hard to resolve this - just port the code over from that.

**Comment by sbulen** [comment]
At one point I had logged an internal documentation suggestion that there may need to be pre-work for custom avatars to work properly following a 2.0 => 2.1 upgrade.  

> (1) Custom avatar folder.  In 2.0, the custom avatar folder is not required, it defaults to being in the attachment directory.  In 2.1, it has a separate directory.  So...  I see 3 issues:
>  - Folks have probably given their custom avatar directory a specific name that is not aligned with where 2.1 expects it (.../custom_avatar)
>  - Some folks may still have avatars within their attachment directory. 
>  - Some folks may & some folks may NOT have done the "hidden" migration step to move existing avatars from the attachment folder to the custom avatar folder.   
> 

So, if we're to try & fix this, I think we need to look at the issue holistically - all 3 potential interrelated concerns above.  

My initial thought was to deal with all of this via instructions & pre-work before the upgrade.

**Comment by sbulen** [comment]
This has been biting others:
https://www.simplemachines.org/community/index.php?topic=559400.0

**Comment by Oldiesmann** [comment]
I still think all we really need to do is determine whether there are avatars in the attachment directory based on the 2.0 setting and move them if necessary. The directory name should carry over from 2.0 if they've chosen something other than "custom_avatar".

**Comment by sbulen** [comment]
Do we need to move the contents of custom_avatar (from the 2.1 install directory) over to the user-specified folder?  There's a couple files in there.  Strange things happen if the blank isn't there.

**Comment by sbulen** [comment]
Since this isn't n rc1, I guess we need this code in 2.1 now...

**Comment by Sesquipedalian** [comment]
I'm not sure why I marked this for RC2 when it is obviously an RC1 issue. 

~Fortunately, this should be very easy to fix. Almost all the code to handle this is already written. The only problem is that the upgrader currently only checks for 2.0's `$modSettings['custom_avatar_dir']` and forgets to check for `$modSettings['avatar_directory']`. A quick update to add a check for both should solve the issue with very little muss or fuss.~

~(Sidebar: I don't really understand the purpose of having both `$modSettings['avatar_directory']` and `$modSettings['custom_avatar_dir']` in 2.0.)~

Disregard previous statements. I'm still looking into it...

**Comment by Sesquipedalian** [comment]
I figured it out. The problem occurred when someone went to the index.php after the upgrade files had been put into place but before the upgrader had been run. 

Specifically, when index.php was hit, some code in `reloadSettings()` would blindly try to decode a plain string value for `$modSettings['attachmentUploadDir']` as if it were JSON, which would of course produce an empty array. Then that empty array would be encoded back to JSON and saved. This meant that later when the upgrader tried to update settings related to attachments and avatars, there would be no attachment directory defined, and so trying to move them around, rename them, etc., would fail.

**Comment by sbulen** [comment]
What you're describing sounds like a completely different issue than the reported one to me...

**Comment by Sesquipedalian** [comment]
Sorry, I suppose I wasn't clear enough. I was describing what caused the following to happen:
> the avatars are not moved over to the custom_avatars directory during the upgrade

Failure to move the image files to the correct directory during upgrade is the cause of everything else described in this issue, and the cause of that failure is what I was describing in my previous comment. 

I agree that it _sounds_ like a completely different issue. One wouldn't expect some bad code in index.php to mess up the upgrader, after all. But the reason it sounds unconnected is the same reason why I spent hours like this before I could figure what the heck was causing the avatars not to be moved correctly.
![394898c0-f44a-45b7-aec5-200cd496a031](https://user-images.githubusercontent.com/191731/51886232-25318600-234c-11e9-96e8-d6d1c05e094e.png)
It's an obscure and weird bug.

**Comment by sbulen** [comment]
When I reproduced your bug, no attachments were displayed properly.  Because the attachment folder name was wrong.  Your bug affected more than just custom_avatars....

Anyway, for me, the point is moot because in my testing tonight, I could not get the migration of custom avatars from the attachments folder to fail.  I tried a variety of configs & couldn't get it to fail.  So, at the moment, I think we're good!
